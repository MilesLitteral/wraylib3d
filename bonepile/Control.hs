{-# LANGUAGE AllowAmbiguousTypes #-}

module Manifest.Control where

import Manifest.Control.NetworkManager
import Manifest.Control.UDP
