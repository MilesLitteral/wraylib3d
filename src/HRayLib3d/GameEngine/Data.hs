module HRayLib3d.GameEngine.Data (
        -- module HRayLib3d.GameEngine.Data.BSP,
        -- module HRayLib3d.GameEngine.Data.GameCharacter,
        -- module HRayLib3d.GameEngine.Data.GLB,
        -- module HRayLib3d.GameEngine.Data.GLTF,
        -- module HRayLib3d.GameEngine.Data.Material,
        -- module HRayLib3d.GameEngine.Data.MD3,
        -- module HRayLib3d.GameEngine.Data.OBJ
    ) where

    -- import HRayLib3d.GameEngine.Data.BSP
    -- import HRayLib3d.GameEngine.Data.GameCharacter
    -- import HRayLib3d.GameEngine.Data.GLB
    -- import HRayLib3d.GameEngine.Data.GLTF
    -- import HRayLib3d.GameEngine.Data.Material
    -- import HRayLib3d.GameEngine.Data.MD3
    -- import HRayLib3d.GameEngine.Data.OBJ