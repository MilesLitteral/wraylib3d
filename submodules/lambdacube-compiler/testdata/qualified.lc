{-# LANGUAGE NoImplicitPrelude #-}

import qualified Prelude as P
import qualified Prelude hiding (a,b) as PH
import qualified Prelude (c,e)
-- import qualified Prelude (c,e) hiding (a,b) as PH -- fails
