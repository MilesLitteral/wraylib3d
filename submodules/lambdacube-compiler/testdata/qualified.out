main is not found
------------ desugared source code
------------ core code