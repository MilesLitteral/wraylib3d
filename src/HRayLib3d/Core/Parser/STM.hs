module HRayLib3d.Core.Parser.STM (parseScriptSTM) where

-- Compile with -threaded
import System.IO ()
import Control.Concurrent ( forkIO )
import Control.Concurrent.STM ( atomically, retry, STM, TVar, newTVarIO, readTVar, writeTVar )
import qualified Text.XML.Light.Input  as XML
import qualified Text.XML.Light.Lexer  as XML
import qualified Text.XML.Light.Types  as XML

-- ([Sum of Loaded XML], <Number of finished transactions>)
type Result = TVar ([XML.Content], Int)

-- Adds x to result and increments the number of finished transactions
addContentToResult :: Result -> XML.Content -> STM ()
addContentToResult result x = do
  (resContent, endCtr) <- readTVar result
  writeTVar result (resContent <> [x], endCtr+1)

-- Waits for the number of finished transactions to reach a limit
-- Then returns the sum of the result
waitForParse :: Result -> Int -> STM [XML.Content]
waitForParse result limit = do
  (resContent, endCtr) <- readTVar result
  if endCtr < limit
    then retry
    else return resContent

parseScriptSTM :: XML.XmlSource s => s -> IO ()
parseScriptSTM source = do
  let n  = 100 :: Int                                       -- Number of threads to spawn    
  result <- newTVarIO ([], 0)                               -- Set up TVar
  mapM_ (forkIO . atomically . addContentToResult result) $ XML.parseXML source--[1..n]   -- Spawn threads
  resSum <- atomically $ waitForParse result n              -- Wait for threads to finish and get sum
  putStrLn $ "Sum [1..n] = " ++ show resSum                 -- Print sum