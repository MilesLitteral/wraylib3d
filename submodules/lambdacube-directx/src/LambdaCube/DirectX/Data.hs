module LambdaCube.DirectX.Data where

import Control.Applicative
import Control.Monad
import Data.IORef
import Data.List as L
import Data.Maybe
import Foreign 
--import qualified Data.IntMap as IM
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Vector as V
import qualified Data.Vector.Storable as SV

--import Control.DeepSeq

import Data.Word
import Codec.Picture
import Codec.Picture.Types

import Graphics.D3D11.Interface --import Graphics.GL.Core33 
import LambdaCube.DirectX.Type
import LambdaCube.DirectX.Util

-- Buffer
disposeBuffer :: ID3D11Buffer -> IO ()
disposeBuffer (Buffer _ bo) = withArray [bo] $ glDeleteBuffers 1

compileBuffer :: [SimpleVertex] -> IO ID3D11Buffer
compileBuffer arrs = do
    let bd = D3D11BufferDesc
        { byteWidth = fromIntegral $ 8 * sizeOf (undefined :: SimpleVertex)
        , usage = D3D11UsageDefault
        , bindFlags = d3d11BindFlags [D3D11BindVertexBuffer]
        , cpuAccessFlags = 0
        , miscFlags = 0
        , structureByteStride = 0 }
    buffer <- createBuffer device bd arrs                  
    iaSetVertexBuffers deviceContext 0 [(buffer, fromIntegral $ sizeOf (undefined :: SimpleVertex), 0)]
    return $! ID3D11Buffer buffer

-- updateBuffer :: ID3D11Buffer -> [(Int,Array)] -> IO ()
-- updateBuffer (Buffer arrDescs bo) arrs = do
--     glBindBuffer GL_ARRAY_BUFFER bo
--     forM arrs $ \(i,Array arrType cnt setter) -> do
--         let ArrayDesc ty len offset size = arrDescs V.! i
--         when (ty == arrType && cnt == len) $
--             setter $! glBufferSubData GL_ARRAY_BUFFER (fromIntegral offset) (fromIntegral size)
--     glBindBuffer GL_ARRAY_BUFFER 0

-- bufferSize :: ID3D11Buffer -> Int
-- bufferSize = V.length . bufArrays

-- arraySize :: ID3D11Buffer -> Int -> Int
-- arraySize buf arrIdx = arrLength $! bufArrays buf V.! arrIdx

-- arrayType :: ID3D11Buffer -> Int -> ArrayType
-- arrayType buf arrIdx = arrType $! bufArrays buf V.! arrIdx

-- Texture
disposeTexture :: TextureData -> IO ()
disposeTexture (TextureData to) = withArray [to] $ glDeleteTextures 1

-- FIXME: Temporary implemenation
uploadTexture2DToGPU :: DynamicImage -> IO TextureData
uploadTexture2DToGPU = uploadTexture2DToGPU' True False True False

uploadTexture2DToGPU' :: ID3D11Device -> DynamicImage -> IO TextureData
uploadTexture2DToGPU' device bitmap' = do
    let bitmap = case bitmap' of
          ImageRGB8   i@(Image w h _) -> bitmap'
          ImageRGBA8  i@(Image w h _) -> bitmap'
          ImageYCbCr8 i@(Image w h _) -> ImageRGB8 $ convertImage i
          di -> ImageRGBA8 $ convertRGBA8 di
        (width,height) = bitmapSize bitmap
            bitmapSize (ImageRGB8  (Image w h _)) = (w,h)
            bitmapSize (ImageRGBA8 (Image w h _)) = (w,h)
            bitmapSize _ = error "unsupported image type :("
            withBitmap (ImageRGB8  (Image w h v)) f = SV.unsafeWith v $ f (w,h) 3 0
            withBitmap (ImageRGBA8 (Image w h v)) f = SV.unsafeWith v $ f (w,h) 4 0
            withBitmap _ _ = error "unsupported image type :("
        td = D3D11Texture2DDesc
                width
                height
                1
                1
                DxgiFormatD24UnormS8Uint
                (DxgiSampleDesc 1 0)
                D3D11UsageDefault
                (d3d11BindFlags []) --D3D11BindDepthStencil
                0
                0
    to <- createTexture2D device td (SV.toList v :: [Word16])
    return $ TextureData to
