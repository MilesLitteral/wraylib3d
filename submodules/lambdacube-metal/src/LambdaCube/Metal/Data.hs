module LambdaCube.Metal.Data where

import Control.Applicative
import Control.Monad
import Data.IORef
import Data.List as L
import Data.Maybe
import Foreign 
--import qualified Data.IntMap as IM
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Vector as V
import qualified Data.Vector.Storable as SV

--import Control.DeepSeq

import Data.Word
import Codec.Picture
import Codec.Picture.Types

import LambdaCube.Metal.Type
import LambdaCube.Metal.Util
import Graphics.GL.Core33
--import LambdaCube.Metal.Bindings  --Graphics.GL.Core33
--Utility Functions
calcDesc (offset,setters,descs) (Array arrType cnt setter) =
          let size = cnt * sizeOfArrayType arrType
            in (size + offset, (offset, size,setter):setters, ArrayDesc arrType cnt offset size:descs)

-- Buffer
disposeBuffer :: Buffer -> IO ()
disposeBuffer (Buffer _ bo) = withArray [bo] $ resetCommandsInBuffer 1 

compileBuffer :: MTLDevice -> [Array] -> IO Buffer
compileBuffer dev arrs = do
    let (bufSize,arrSetters,arrDescs) = foldl' calcDesc (0,[],[]) arrs
    bo <- alloca $! \pbo -> makeBuffer dev pbo 1 MTLResourceOptionCPUCacheModeDefault >> peek pbo --glGenBuffers
    forM_ arrSetters $! \(offset,size,setter) -> setter $! makeBuffer dev (fromIntegral offset) (fromIntegral size) MTLResourceOptionCPUCacheModeDefault
    return $! Buffer (V.fromList $! reverse arrDescs) bo

updateBuffer :: MTLDevice -> Buffer -> [(Int,Array)] -> IO ()
updateBuffer (Buffer arrDescs bo) arrs = forM arrs $ 
    \(i,Array arrType cnt setter) -> do
        let ArrayDesc ty len offset size = arrDescs V.! i
        when (ty == arrType && cnt == len) $ 
            setter $! makeBuffer dev (fromIntegral offset) (fromIntegral size) MTLResourceOptionCPUCacheModeDefault

bufferSize :: Buffer -> Int
bufferSize = V.length . bufArrays

arraySize :: Buffer -> Int -> Int
arraySize buf arrIdx = arrLength $! bufArrays buf V.! arrIdx

arrayType :: Buffer -> Int -> ArrayType
arrayType buf arrIdx = arrType $! bufArrays buf V.! arrIdx

-- Texture
disposeTexture :: TextureData -> IO ()
disposeTexture (TextureData to) = withArray [to] $ metalDeleteTextures --glDeleteTextures 1

-- FIXME: Temporary implemenation
uploadTexture2DToGPU :: DynamicImage -> IO TextureData
uploadTexture2DToGPU = uploadTexture2DToGPU' True False True False

uploadTexture2DToGPU' :: Bool -> DynamicImage -> IO TextureData
uploadTexture2DToGPU' isMip bitmap' = do
    let bitmap = case bitmap' of
                    ImageRGB8   i@(Image w h _) -> bitmap'
                    ImageRGBA8  i@(Image w h _) -> bitmap'
                    ImageYCbCr8 i@(Image w h _) -> ImageRGB8 $ convertImage i
                    di -> ImageRGBA8 $ convertRGBA8 di
        (width,height) = bitmapSize bitmap
                         bitmapSize (ImageRGB8  (Image w h _)) = (w,h)
                         bitmapSize (ImageRGBA8 (Image w h _)) = (w,h)
                         bitmapSize _ = error "unsupported image type :("
                         withBitmap (ImageRGB8 (Image w h v))  f = SV.unsafeWith v $ f (w,h) 3 0
                         withBitmap (ImageRGBA8 (Image w h v)) f = SV.unsafeWith v $ f (w,h) 4 0
                         withBitmap _ _ = error "unsupported image type :("
        to <- alloca $! \pto -> newTextureWithDescriptor pto >> MTLTextureDescriptor PixelFormatBGRA8Unorm width height --peek pto
        replaceRegion 0 v $ castPtr ptr
    when isMip $ replaceRegion 1 v $ castPtr ptr
    return $ TextureData to
