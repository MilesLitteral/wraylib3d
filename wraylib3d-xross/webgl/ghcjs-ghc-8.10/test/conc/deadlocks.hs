{-# LANGUAGE ForeignFunctionInterface, JavaScriptFFI, ScopedTypeVariables #-}
{- 
   some tests for the storage manager

   checks that deadlocks are detected and dealt with properly
-}
module Main where

import System.Mem
import System.Mem.Weak
import Control.Concurrent
import Control.Concurrent.MVar
import Control.Concurrent.STM
import Control.Monad
import Data.Maybe

import qualified Control.Exception as E

delay :: IO ()
delay = threadDelay 4000000 -- long enough for GC to kick in

foreign import javascript interruptible "setTimeout($c, 4000);" js_delay :: IO ()

expectException :: String -> IO () -> IO ()
expectException xs m = (m >> putStrLn ("ERROR: expected exception: " ++ xs))
                       `E.catch`
                       \(_::E.SomeException) -> putStrLn ("OK: " ++ xs)

printException :: String -> IO () -> IO ()
printException xs m = m `E.catch` \(e::E.SomeException) -> putStrLn (xs ++ ": " ++ show e)

main :: IO ()
main = do
  testFFI
  testMVar
  testSTM
  testWeak

testFFI :: IO ()
testFFI = putStrLn "testFFI" >> sequence_ [test1]
  where
    test1 = do
      mv1 <- newEmptyMVar
      forkIO $ js_delay >> putMVar mv1 "FFI test 1 OK"
      putStrLn =<< takeMVar mv1

testMVar :: IO ()
testMVar = putStrLn "testMVar" >> sequence_ [test1,test2,test3]
  where
    -- unreachable MVar
    test1 = expectException "MVar test 1" $ do 
      mv1 <- newEmptyMVar
      void (takeMVar mv1)
    test2 = do
      mv1 <- newEmptyMVar
      mv2 <- newEmptyMVar
      let thread a b = forkIO $ printException "MVar test 2" (void $ takeMVar a) >> putMVar b () >> putStrLn "OK"
      -- one of these should get an exception, both should finish and print OK
      thread mv1 mv2
      thread mv2 mv1
      delay
    test3 = do
      let chain mv1 0 = return mv1
          chain mv1 n = do
            mv2 <- newEmptyMVar
            forkIO $ takeMVar mv1 >>= \v -> threadDelay 1000000 >> putMVar mv2 v
            return mv2
      mv1 <- newEmptyMVar
      mv2 <- chain mv1 50
      delay
      putMVar mv1 "MVar test3: OK"
      putStrLn =<< takeMVar mv2

testWeak :: IO ()
testWeak = putStrLn "testWeak" >> sequence_ [test1]
  where
    test1 = do
      mv <- newEmptyMVar
      wmv1 <- mkWeakMVar mv (putStrLn "Weak test 1 finalizer")
      putStrLn "Weak test 1"
      forkIO $ delay >> putMVar mv "OK"
      print . isJust =<< deRefWeak wmv1
      putStrLn =<< takeMVar mv
      delay
      print . isJust =<< deRefWeak wmv1


testSTM :: IO ()
testSTM = putStrLn "testSTM" >> sequence_ [test1]
  where
    test1 = expectException "STM test 1" $ do
      tv1 <- atomically (newTVar (0::Int))
      forkIO $ atomically (readTVar tv1 >>= \v -> if v < 1 then retry else writeTVar tv1 (v+1))
      delay
      atomically (modifyTVar tv1 (+1))
      print =<< atomically (readTVar tv1 >>= \v -> if v < 2 then retry else return v)
      atomically (readTVar tv1 >>= \v -> if v < 3 then retry else return ())
      putStrLn "ERROR"

