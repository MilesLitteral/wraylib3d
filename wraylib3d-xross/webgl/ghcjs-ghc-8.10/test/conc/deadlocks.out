testFFI
FFI test 1 OK
testMVar
OK: MVar test 1
MVar test 2: thread blocked indefinitely in an MVar operation
OK
OK
MVar test3: OK
testSTM
2
OK: STM test 1
testWeak
Weak test 1
True
OK
Weak test 1 finalizer
False

